File: src/day8.py
from collections import Counter
segments = {3:7, 4:4, 2:1, 7:8}
count = Counter()
map = {}
for line in open('input/2021/test8.txt'):
# for line in open('input/2021/day8.txt'):
    inp, out = line.split('|')
    for o in out.split():
        try:
            m = segments[len(o)]
            count[m] += 1
        except KeyError:
            pass
print(sum(count.values()))

total = 0
# for line in open('input/2021/test8.txt'):
for line in open('input/2021/day8.txt'):
    dd = {}
    inp, out = line.split('|')
    print(inp, flush=True)
    undecoded = set(inp.split())
    while undecoded:
        drop = set()
        for sig in undecoded:
            ssig = set(sig)
            try:
                if len(sig) in segments:
                    dd[segments[len(sig)]] = ssig
                    drop.add(sig)
                    dd['a'] = (dd[8]-dd[4]).intersection(dd[7])
                elif len(sig) == 6:
                    x = dd[8] - ssig - dd[1]
                    if not x:
                        dd[6] = ssig
                        drop.add(sig)
                        dd['c'] = dd[8] - dd[6]
                        dd['f'] = dd[6].intersection(dd[1] - dd['c'])
                    else:
                        x = dd[8] - ssig - dd[4]
                        if x:
                            dd[9] = ssig
                            dd['e'] = dd[8] - dd[9]
                        else:
                            dd[0] = ssig
                            dd['d'] = dd[8] - dd[0]
                        drop.add(sig)
                else: #2, 3, 5
                    if ssig.intersection(dd[1]) == dd[1]:
                        dd[3] = ssig
                        dd['b'] = dd[8] - dd[3] - dd['e']
                        drop.add(sig)
                    elif ssig.intersection(dd['b']):
                        dd[5] = ssig
                        drop.add(sig)
                    else:
                        dd[2] = ssig
                        drop.add(sig)
                    
                     
            except KeyError:
                continue
        undecoded -= drop
    num = ''
    de = {}
    for k, v in dd.items():
        de[tuple(sorted(v))] = str(k)
    for o in out.split():
        num += de[tuple(list(sorted(o)))]
    total += int(num)
print(total)
    # print(dd)



# data = [int(x) for x in open('input/2021/day8.txt').read().split(',')]
#   0:      1:      2:      3:      4:
#  aaaa    ....    aaaa    aaaa    ....
# b    c  .    c  .    c  .    c  b    c
# b    c  .    c  .    c  .    c  b    c
#  ....    ....    dddd    dddd    dddd
# e    f  .    f  e    .  .    f  .    f
# e    f  .    f  e    .  .    f  .    f
#  gggg    ....    gggg    gggg    ....

#   5:      6:      7:      8:      9:
#  aaaa    aaaa    aaaa    aaaa    aaaa
# b    .  b    .  .    c  b    c  b    c
# b    .  b    .  .    c  b    c  b    c
#  dddd    dddd    ....    dddd    dddd
# .    f  e    f  .    f  e    f  .    f
# .    f  e    f  .    f  e    f  .    f
#  gggg    gggg    ....    gggg    gggg

File: src/day_8.rs
use std::collections::HashSet;
type Item = HashSet<char>;
type Items = Vec<Item>;
type Input<'a> = Vec<(Items, Items)>;

pub fn input_generator(input: &str) -> Input {
    input.lines().map(|line| {
        let (signals, digits) = line.split_once('|').unwrap();
        let signals: Items = signals.split_whitespace().map(|x| x.chars().collect()).collect();
        let digits:  Items = digits.split_whitespace().map(|x| x.chars().collect()).collect();
        (signals, digits)
    }).collect()
}

/// Ignore the signals for Part 1 - just need to count easily identifiable digits
pub fn part1(input: &Input) -> usize {
    input
        .iter()
        .map(|(_, digits)| {
            digits.iter().filter(|digit| [2, 3, 4, 7].contains(&digit.len())).count()
        }).sum()
}

/// Identify the signals by deduction, then decode the digits
pub fn part2(input: &Input) -> usize {
    let mut d2s: Vec<Item> = (0..10).map(|_| HashSet::<char>::new()).collect();
    let mut total: usize = 0;
    for (sigs, digits) in input {
        d2s[1] = sigs.iter().filter(|s| s.len()==2).next().unwrap().clone();
        d2s[8] = sigs.iter().filter(|s| s.len()==7).next().unwrap().clone();
        d2s[7] = sigs.iter().filter(|s| s.len()==3).next().unwrap().clone();
        d2s[4] = sigs.iter().filter(|s| s.len()==4).next().unwrap().clone();

        d2s[9] = sigs.iter().filter(|s| s.len()==6 && s.is_superset(&d2s[4])).next().unwrap().clone();
        d2s[6] = sigs.iter().filter(|s| s.len()==6 && s.intersection(&d2s[1]).copied().collect::<Item>().len()==1).next().unwrap().clone();
        d2s[0] = sigs.iter().filter(|s| s.len()==6 && *s != &d2s[9] && *s != &d2s[6]).next().unwrap().clone();

        d2s[3] = sigs.iter().filter(|s| s.len()==5 && s.is_superset(&d2s[1])).next().unwrap().clone();
        d2s[2] = sigs.iter().filter(|s| s.len()==5 && s.intersection(&d2s[4]).copied().collect::<Item>().len()==2).next().unwrap().clone();
        d2s[5] = sigs.iter().filter(|s| s.len()==5 && *s != &d2s[3] && *s != &d2s[2]).next().unwrap().clone();

        let num: String = digits.iter().map(|digit| d2s.iter().position(|sig| sig==digit).unwrap().to_string()).collect();
        total += num.parse::<usize>().unwrap();
    }
    total
}
